cpp/ql/src/semmle/code/cpp/controlflow/internal/CFG.qll

```diff
@@ -983,36 +981,65 @@ private predicate subEdgeIncludingDestructors(Pos p1, Node n1, Node n2, Pos p2)
-    exists(int i |
-      p1.nodeAt(n1, getDestructorCallAfterNode(n, i).getAccess()) and
-      p2.nodeAt(n2, getDestructorCallAfterNode(n, i))
-    )
+    // before destructors -> access(max)
+    exists(int maxCallIndex |
+      maxCallIndex = max(int i | exists(getSynthesisedDestructorCallAfterNode(n, i))) and
+      p1.nodeBeforeDestructors(n1, n) and
+      p2.nodeAt(n2, getSynthesisedDestructorCallVariableAccessAfterNode(n, maxCallIndex)))
     or
-    // call(i) -> access(i+1)
+    // call(i+1) -> access(i)
     exists(int i |
-      p1.nodeAt(n1, getDestructorCallAfterNode(n, i)) and
-      p2.nodeAt(n2, getDestructorCallAfterNode(n, i + 1).getAccess())
+      p1.nodeAt(n1, getSynthesisedDestructorCallAfterNode(n, i + 1)) and
+      p2.nodeAt(n2, getSynthesisedDestructorCallVariableAccessAfterNode(n, i))
     )
     or
-    // call(max) -> after destructors end
-    exists(int maxCallIndex |
-      maxCallIndex = max(int i | exists(getDestructorCallAfterNode(n, i))) and
-      p1.nodeAt(n1, getDestructorCallAfterNode(n, maxCallIndex)) and
-      p2.nodeAfterDestructors(n2, n)
-    )
+    // call(0) -> after destructors end
+    p1.nodeAt(n1, getSynthesisedDestructorCallAfterNode(n, 0)) and
+    p2.nodeAfterDestructors(n2, n)
   )
 }
 
+/**
+ * Gets the `index`'th synthetic destructor call that should follow `node`. The
+ * exact placement of that call in the CFG depends on the type of `node` as
+ * follows:
+ *
+ * - `Block`: after ordinary control flow falls off the end of the block
+ *   without jumps or exceptions.
+ * - `ReturnStmt`: After the statement itself or after its operand (if
+ *   present).
+ * - `ThrowExpr`: After the `throw` expression or after its operand (if
+ *   present).
+ * - `JumpStmt` (`BreakStmt`, `ContinueStmt`, `GotoStmt`): after the statement.
+ * - A `ForStmt`, `WhileStmt`, `SwitchStmt`, or `IfStmt`: after control flow
+ *   falls off the end of the statement without jumping. Destruction can occur
+ *   here for `for`-loops that have an initializer (`for (C x = a; ...; ...)`)
+ *   and for statements whose condition is a `ConditionDeclExpr`
+ *   (`if (C x = a)`).
+ * - The `getUpdate()` of a `ForStmt`: after the `getUpdate()` expression. This
+ *   can happen when the condition is a `ConditionDeclExpr`
+ * - `Handler`: On the edge out of the `Handler` for the case where the
+ *   exception was not matched and is propagated to the next handler or
+ *   function exit point.
+ * - `MicrosoftTryExceptStmt`: After the false-edge out of the `e` in
+ *   `__except(e)`, before propagating the exception up to the next handler or
+ *   function exit point.
+ * - `MicrosoftTryFinallyStmt`: On the edge following the `__finally` block for
+ *   the case where an exception was thrown and needs to be propagated.
+ */
+DestructorCall getSynthesisedDestructorCallAfterNode(Node n, int i) {
+  synthetic_destructor_call(n, i, result)
+}
+
+VariableAccess getSynthesisedDestructorCallVariableAccessAfterNode(Node n, int i) {
```

Comment:
Please add QLDoc -- or just inline the predicate if you can write `getChild(-1)` in a more self-explanatory way.